File: 0129-sum-root-to-leaf-numbers/0129-sum-root-to-leaf-numbers.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def sumNumbers(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        return self.sumNumbersHelper(root, 0)

    def sumNumbersHelper(self, node, currentSum):
        if not node:
            return 0

        currentSum = currentSum * 10 + node.val

        if not node.left and not node.right:
            return currentSum

        return self.sumNumbersHelper(node.left, currentSum) + self.sumNumbersHelper(node.right, currentSum)
        

File: 0129-sum-root-to-leaf-numbers/0129-sum-root-to-leaf-numbers.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int sumNumbers(TreeNode root) {
        return sumNumbersHelper(root, 0);
    }

    private int sumNumbersHelper(TreeNode node, int currentSum) {
        if (node == null) {
            return 0;
        }

        // 현재 경로의 숫자 업데이트
        currentSum = currentSum * 10 + node.val;

        // 리프 노드인 경우, 현재 경로의 숫자를 반환
        if (node.left == null && node.right == null) {
            return currentSum;
        }

        // 좌우 서브트리의 합을 재귀적으로 계산하여 더함
        return sumNumbersHelper(node.left, currentSum) + sumNumbersHelper(node.right, currentSum);
    }
}